Add third-party skeletal features

karthikprabhuvinod opened the pull request:
This PR starts the third-party repo-add, delete, list feature
for swupd. This PR also add support for configure flag
to build swupd optionally, off by default.

**Repo Add:**
 sudo swupd 3rd-party add <repo_name> \<url\>
**Repo Remove:**
 sudo swupd 3rd-party remove <repo_name>
**Repos List:**
 sudo swupd 3rd-party list

Signed-off-by: Karthik Prabhu Vinod <karthik.prabhu.vinod@intel.com>

karthikprabhuvinod commented:
@otaviobp : I have a working third_party_repo PR...Tests are still pending and PR needs a bit of finishing. At this moment, could you  take a look from an early feedback & feature standpoint ?

otaviobp commented on Makefile.am:
```diff
@@ -32,6 +32,11 @@ bin_PROGRAMS = swupd
+	src/3rd_party_add.c \
+	src/3rd_party.c \
+	src/3rd_party_helpers.c \
+        src/3rd_party_list.c \
```
mix tab/spaces?

karthikprabhuvinod replied:
Makes sense will change in next push

otaviobp commented on configure.ac:
```diff
@@ -72,6 +72,25 @@ AS_IF(
 )
 
+AC_ARG_ENABLE(
+  [thirdparty],
```
I prefer third-party, but I'm ok with all together too.

otaviobp commented on configure.ac:
```diff
@@ -72,6 +72,25 @@ AS_IF(
+  [THIRDPARTY="no"]
+)
+
+AH_TEMPLATE([THIRDPARTY_REPO_PREFIX], [prefix path for third party repos])
```
++ for having a setting for REPO_PREFIX

otaviobp commented on src/3rd_party.c:
```diff
@@ -0,0 +1,173 @@
+
+void get_repo_path(char **repo_path, char *repo_name)
+{
+	string_or_die(repo_path, "%s%s", THIRDPARTY_REPO_PREFIX, repo_name);
```
Use sys_path_join here to make sure we are going to have a '/' separator

karthikprabhuvinod replied:
Makes sense will change in next push

otaviobp commented on src/3rd_party.c:
```diff
@@ -0,0 +1,173 @@
+	global_print_help();
+
+	print("Options:\n");
+	print("   --add <repo_name>		Add third party repo\n");
```
I don't think it's a good idea to use flags for add, list remove. If you use a flag we need to protect against multiple flags used together and this will become hard after adding bundle-add and other commands

karthikprabhuvinod replied:
> I don't think it's a good idea to use flags for add, list remove. If you use a flag we need to protect against multiple flags used together and this will become hard after adding bundle-add and other commands

I was thinking about that....I need to have add, list, remove as sub_commands like add in git remote add

karthikprabhuvinod replied:
Makes sense will change in next push

otaviobp commented on src/3rd_party.c:
```diff
@@ -0,0 +1,173 @@
+
+enum swupd_code third_party_main(int UNUSED_PARAM argc, char UNUSED_PARAM **argv)
+{
+	info("This feature is disabled at compile time\n");
```
It's better to avoid a print here so for users this feature won't exist before it's released officially.

karthikprabhuvinod replied:
Makes sense will change in next push

otaviobp commented on src/3rd_party.h:
```diff
@@ -0,0 +1,8 @@
+#ifndef __THIRD_PARTY_H
+#define __THIRD_PARTY_H
+
```
I would add the `#ifdef THIRDPARTY` here and never call those functions if it's false.

otaviobp replied:
And follow the standard in other header files to create this one. The define name is different and there's `extern "C" `missing

karthikprabhuvinod replied:
> I would add the `#ifdef THIRDPARTY` here and never call those functions if it's false.

I think we are using the same pattern using compile time flag for signatures...like in c src/signature.c

karthikprabhuvinod replied:
> And follow the standard in other header files to create this one. The define name is different and there's `extern "C" `missing

Makes sense will change in next push

otaviobp replied:
I mixed 2 comments in same line and I realized now that it's confusing.
 - The #ifdef style is good the way it is, like the signature as you said. But If you just ignore all file it's going to be easier to code and I'm ok with that. But if you want to leave like it is, it's fine too.
 - The style I was talking about is the header file. But I realized that there's not a very well established style. each file does it in a different way. So the only thing really missing is the extern.

karthikprabhuvinod replied:
For point1 I will try to do what you said. I agree with you
For point2, will make sure headers have extern c

karthikprabhuvinod replied:
Makes sense will change in next push

otaviobp commented on swupd.bash:
```diff
@@ -39,7 +39,7 @@ _swupd()
 		opts="$global "
 		break;;
-		("update")
+	    ("update")
```
tab/spaces?

karthikprabhuvinod replied:
Makes sense will change in next push

otaviobp commented on src/3rd_party.c:
```diff
@@ -0,0 +1,173 @@
+
+	if (!cmdline_option_repo) {
+		if (cmdline_option_add || cmdline_option_remove) {
+			error("The --repo option is a required argument\n\n");
```
same for --repo. I prefer:

`swupd 3rd-party add repo_name url`

karthikprabhuvinod replied:
Actually I was thinking this should be an option for sure. Usually command are actions.

It works perfectly in our case as then we can make it required for remove, add and not have it for list.

Let me know what you feel

otaviobp commented on src/3rd_party_add.c:
```diff
@@ -0,0 +1,59 @@
+
+int third_party_add(char *repo_path)
+{
+	if (repo_exists(repo_path)) {
```
We need to have a file in /var/lib/swupd to track the 3rd-party repositories and the corresponding url, so checking for the dir is not enough. And if the dir is completly removed we should be able to restore that.

karthikprabhuvinod replied:
Will address this in a follow-on PR as discussed

otaviobp commented on src/main.c:
```diff
@@ -69,6 +69,7 @@ static struct subcmd commands[] = {
 	{ "clean", "Clean cached files", clean_main },
 	{ "hashdump", "Dump the HMAC hash of a file", hashdump_main },
 	{ "verify", "NOTE: this command has been superseded, please use \"swupd diagnose\" instead", diagnose_main },
+	{ "3rd-party", "Indicate Swupd to use user's third party repo", third_party_main },
```
And use the ifdef here too. I forgot to mention that in my first review

karthikprabhuvinod replied:
Ahh yes.. will do

karthikprabhuvinod replied:
Makes sense will change in next push

karthikprabhuvinod commented:
@otaviobp 
I have addressed all your comments with remarks like "Makes sense will change in next push" except two i guess.

With this PR, I made sure the skeleton exists, The add, delete list command lines parsers are created. But they don't do much. 

Please could give it a second quick look? Anything you want to be removed for now?

otaviobp commented on src/3rd_party.h:
```diff
@@ -0,0 +1,44 @@
+ * @return 0 on success, SWUPD_THIRD_PARTY_REPO_EXISTS if it already exists
+ * and negative numbers on errors
+ */
+enum swupd_code third_party_add(int argc, char **argv);
```
Rename all functions that are a main to add a _main suffix. This one should be `third_party_add_main`

karthikprabhuvinod replied:
Makes sense will change in next push

otaviobp commented on src/3rd_party.c:
```diff
@@ -0,0 +1,136 @@
+	{ 0 }
+};
+
+static int subcmd_index(char *arg)
```
Instead of copying, export this function in swupd_internal.h and use it in 3rd_party

karthikprabhuvinod replied:
@otaviobp This command is slightly diffferent than the one we used in main.c since we dont have to check if it is a verify command etc. Should I still use that same function? Also I will need to pass commands args to it since commands are different from main.c, add.c remove.c

otaviobp replied:
take a look at  #1180
And you can get that as a parameters. You can rename that if you find a better name and add a comment explaining what the function is doing.

karthikprabhuvinod replied:
Okay Cool I will move to swupd_internal.h as a inline and include it in main.c and 3rd-party.c

karthikprabhuvinod replied:
Makes sense will change in next push

karthikprabhuvinod commented on Makefile.am:
```diff
@@ -32,6 +32,10 @@ bin_PROGRAMS = swupd
 AM_CPPFLAGS = $(AM_CFLAGS) $(libarchive_CFLAGS)
 
 swupd_SOURCES = \
+	src/3rd_party_add.c \
```
Makes sense will change in next push

karthikprabhuvinod commented on src/3rd_party.h:
```diff
@@ -0,0 +1,44 @@
+extern "C" {
+#endif
+
+#include "swupd.h"
```
Makes sense will change in next push

karthikprabhuvinod commented on src/3rd_party.h:
```diff
@@ -0,0 +1,44 @@
+#include "swupd.h"
+
+/**
+ * @brief Creates a repo folder under THIRDPARTY_REPO_PREFIX 
```
Makes sense will change in next push

karthikprabhuvinod commented on src/3rd_party_add.c:
```diff
@@ -0,0 +1,120 @@
+static int create_repo_dir(char UNUSED_PARAM *repo_path)
+{
+	//repo_path doesn't exist
+	/*
```
Makes sense will change in next push

karthikprabhuvinod commented on src/3rd_party_internal.h:
```diff
@@ -0,0 +1,109 @@
+	char *tracking_file;
+	int ret;
+	tracking_dir = sys_path_join(globals.path_prefix, globals.state_dir);
+	tracking_file = sys_path_join(tracking_dir, "repo_tracker.ini");
```
Makes sense will change in next push

karthikprabhuvinod commented on src/swupd_exit_codes.h:
```diff
@@ -47,6 +47,8 @@ enum swupd_code {
 	SWUPD_OUT_OF_MEMORY_ERROR,	     /* 34 swupd ran out of memory */
 	SWUPD_VERIFY_FAILED,		     /* 35 verify could not fix/replace/delete one or more files */
 	SWUPD_INVALID_BINARY,		     /* 36 binary to be executed is missing or invalid */
+	SWUPD_THIRD_PARTY_REPO_EXISTS,	     /* 37 third party repo already exists, error for adding */
```
Makes sense will change in next push

karthikprabhuvinod commented on src/3rd_party_internal.h:
```diff
@@ -0,0 +1,109 @@
+ */
+static inline bool repo_exists(char *repo_path)
+{
+	struct stat sb;
```
Makes sense will change in next push

karthikprabhuvinod commented on src/3rd_party_internal.h:
```diff
@@ -0,0 +1,109 @@
+	return false;
+}
+
+static int create_tracking_file(char *tracking_file)
```
@otaviobp : for me, a name like create_repo makes more sense with something like creating a repo directory. Here we are talking about creating a tracking file.

otaviobp replied:
So probably I didn't understand the usage of this function. What's a tracking file?

karthikprabhuvinod replied:
Its the repo.ini file where write what repo it is and all the bundles under it

karthikprabhuvinod replied:
Makes sense will change in next push

castulo commented on src/swupd_internal.h:
```diff
@@ -17,5 +17,36 @@ extern enum swupd_code install_main(int argc, char **argv);
+	int i = 0;
+	size_t input_len;
+	size_t cmd_len;
+
```
why did you remove this from the function?
```
if (strcmp(arg, "verify") == 0) {		
 	verify_set_command_verify(true); // set to true so we know the "verify" command was used		
 	arg = "diagnose";		
 }
```

otaviobp replied:
see #1180

castulo replied:
just did, that explains it, thanks

karthikprabhuvinod replied:
Marking this as resolved.

castulo commented on swupd.zsh:
```diff
@@ -175,14 +175,41 @@ if [[ -n "$state" ]]; then
         "mirror:Configure mirror url for swupd content"
         "clean:Clean cached files"
         "hashdump:Dump the HMAC hash of a file"
+	"3rd-party: Indicate Swupd to use user's third party repo"
```
you seem to be mixing tabs and spaces in these files

otaviobp commented:
> This PR seems to be moving in the right direction but there are still many things that need to be addressed first. I added comments with all my observations in the PR. I also noted there are no tests whatsoever for this PR, it introduces lots of new functionality so I think having tests is crucial.

This PR isn't really adding a new functionality, so there's no way on creating tests to it. It's just adding the skeleton so we can move forward.

otaviobp commented:
@karthikprabhuvinod Can you remove all functions from this PR that aren't related to the skeleton? And you can add later when you really need them.

karthikprabhuvinod commented on src/3rd_party.c:
```diff
@@ -0,0 +1,110 @@
+#include "3rd_party.h"
+#include "config.h"
+#include "lib/log.h"
+#include "swupd.h"
```
Makes sense will remove it in next push

karthikprabhuvinod commented on src/3rd_party_add.c:
```diff
@@ -0,0 +1,119 @@
+#include "config.h"
+#include "lib/log.h"
+#include "swupd.h"
+#include "sys/stat.h"
```
There are some others which are redundant as well. Makes sense will remove it in next push

karthikprabhuvinod commented on src/3rd_party_add.c:
```diff
@@ -0,0 +1,119 @@
+	} else if (argc == 2) {
+		error("The positional args: repo-URL is missing\n\n");
+		return false;
+	}
```
Good point.  but Actually am not expecting any more args as such for these commands.

karthikprabhuvinod replied:
Makes sense.. will change it in next push

karthikprabhuvinod commented on src/3rd_party_add.c:
```diff
@@ -0,0 +1,119 @@
+	ret = create_repo_dir(repo_path);
+
+	if (!ret) {
+		info("Third party repo successfully created\n");
```
@castulo Very sharp observation. Yes will change it.

karthikprabhuvinod commented on src/3rd_party_add.c:
```diff
@@ -0,0 +1,119 @@
+	if (!ret) {
+		info("Third party repo successfully created\n");
+		create_repo(argv[1]);
+		goto finish;
```
Makes sense will remove it in next push

karthikprabhuvinod commented on src/3rd_party_add.c:
```diff
@@ -0,0 +1,119 @@
+
+enum swupd_code third_party_add_main(int UNUSED_PARAM argc, char UNUSED_PARAM **argv)
+{
+	return SWUPD_OK;
```
No this one doesnt need a TODO as it is function definition when we enable-third-party is provided to auto-configure

otaviobp replied:
You don't need this function here. REmove it and leave everything related to 3rd party inside the #ifdef

karthikprabhuvinod replied:
Done

karthikprabhuvinod commented on src/3rd_party_internal.h:
```diff
@@ -0,0 +1,98 @@
+#include <stdbool.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <unistd.h>
```
Thanks . Makes sense will remove it in next push

karthikprabhuvinod commented on src/3rd_party_internal.h:
```diff
@@ -0,0 +1,98 @@
+	return SWUPD_OK;
+}
+
+static inline int setup_repo_config()
```
Thanks. Makes sense will remove it in next push

karthikprabhuvinod commented on src/3rd_party_internal.h:
```diff
@@ -0,0 +1,98 @@
+
+	if (create_repo_config(repo_config_file)) {
+		debug("Issue creating repo config file: %s in %s\n", repo_config_file, globals.state_dir);
+		goto finish;
```
Makes sense will remove it in next push

karthikprabhuvinod commented on src/3rd_party_internal.h:
```diff
@@ -0,0 +1,98 @@
+{
+	char *repo_config_file;
+	int ret;
+	repo_config_file = sys_path_join(globals.state_dir, "/3rd_party/repo.ini");
```
Okay. Makes sense will remove it in next push

karthikprabhuvinod commented on src/3rd_party_internal.h:
```diff
@@ -0,0 +1,98 @@
+	return;
+}
+
+static inline char *get_repos()
```
Makes sense will remove it in next push

karthikprabhuvinod commented on src/3rd_party_list.c:
```diff
@@ -0,0 +1,106 @@
+#include "3rd_party_internal.h"
+#include "config.h"
+#include "lib/log.h"
+#include "swupd.h"
```
Makes sense will change it in next push

karthikprabhuvinod commented on src/3rd_party_remove.c:
```diff
@@ -0,0 +1,110 @@
+
+int remove_repo_dir(char UNUSED_PARAM *repo_path)
+{
+	//rm_rf(repo_path);
```
No, will remove and add a TODO

karthikprabhuvinod commented on src/3rd_party_remove.c:
```diff
@@ -0,0 +1,110 @@
+	const int step_in_third_party_remove = 1;
+	char *repo_path = NULL;
+
+	/* there is no need to report in progress for init at this time */
```
Makes sense will change it in next push

karthikprabhuvinod commented on src/3rd_party.h:
```diff
@@ -0,0 +1,44 @@
+ * @return 0 on success, SWUPD_THIRD_PARTY_REPO_EXISTS if it already exists
+ * and negative numbers on errors
+ */
+enum swupd_code third_party_add_main(int argc, char **argv);
```
Thanks for that explanation. Makes sense

karthikprabhuvinod commented:
> @karthikprabhuvinod Can you remove all functions from this PR that aren't related to the skeleton? And you can add later when you really need them.

Okay done

karthikprabhuvinod commented on src/3rd_party_add.c:
```diff
@@ -0,0 +1,119 @@
+	ret = create_repo_dir(repo_path);
+
+	if (!ret) {
+		info("Third party repo successfully created\n");
```
Removed and added a TODO

karthikprabhuvinod commented on src/3rd_party_list.c:
```diff
@@ -0,0 +1,106 @@
+
+static void print_repo_list(char UNUSED_PARAM *repos)
+{
+	return;
```
Removed  and replaced with TODO

karthikprabhuvinod commented on src/3rd_party_remove.c:
```diff
@@ -0,0 +1,110 @@
+		error("The positional args: repo-name is missing\n\n");
+		return false;
+	}
+
```
for above: https://github.com/clearlinux/swupd-client/pull/1161#discussion_r340222876

If think I can do this. I can pre define the no of args expected for each operation (**EXACT_ARG_COUNT**).

Fo eg: for add ...that is 2 (repo-url and repo-name) for remove it is 1(repo-name)

Lets take add as an example:

if (argc - ind) == EXACT_ARG_COUNT

I always check if argc - opind == EXACT_ARG_COUNT..if not fail

otaviobp commented on src/main.c:
```diff
@@ -49,7 +43,7 @@ static enum swupd_code external_search_main(int argc, char **argv)
 // TODO(castulo): remove the superseded command from the help menu by end of November 2019,
 // so it is not visible but the command must remain available in the back so we don't break users
-static struct subcmd commands[] = {
+struct subcmd main_commands[] = {
```
I realized that now. This should remain static.

otaviobp replied:
No. static means that the variable isn't going to be visible outside of this C source file.Using it as a variable is valid and it's ok. Doing that prevents another source file to do this:

```
extern struct subcmd main_commands[];
main_commands = NEW_VALUE;
```

This would change the value of the original main_commands variable

karthikprabhuvinod replied:
@otaviobp Thanks for explaining that. I know why it works...subcmd_index is inline static fucntion and it isnt considered another file.

karthikprabhuvinod replied:
Fixed this in last push

karthikprabhuvinod commented:
> > ignore
> 
> What should I ignore?

I accidently reviewed my own PR..so had to paste some comment

karthikprabhuvinod commented on src/3rd_party.c:
```diff
@@ -0,0 +1,110 @@
+
+#ifdef THIRDPARTY
+
+struct subcmd third_party_commands[] = {
```
same as discussed in
https://github.com/clearlinux/swupd-client/pull/1161#discussion_r340778718

karthikprabhuvinod replied:
Fixed this in last push

karthikprabhuvinod commented:
@otaviobp @castulo Looks like all comments are addressed. I think it looks ready now

castulo commented on src/3rd_party_add.c:
```diff
@@ -0,0 +1,94 @@
+
+	/* Ensure that repo add only expects only two args: repo-name, repo-url */
+	if ((argc - ind) != EXACT_ARG_COUNT) {
+		error("Too many Unexpected args\n\n");
```
nitpick, "Too many Unexpected args" sounds redundant, use "Too many arguments" or "Unexpected arguments", also you are using a capital letter in Unexpected in the middle of the sentence

castulo replied:
Actually to be consistent with other swupd commands it is better to go with `error("unexpected arguments\n\n");`.

karthikprabhuvinod replied:
Sure thing Will correct it in next push

castulo commented on src/3rd_party_list.c:
```diff
@@ -0,0 +1,83 @@
+	}
+
+	if (argc > ind) {
+		error("Too many Unexpected args\n\n");
```
same here -> `error("unexpected arguments\n\n");`

karthikprabhuvinod replied:
Sure thing Will correct it in next push

castulo commented on src/swupd_internal.h:
```diff
@@ -22,6 +25,59 @@ enum swupd_code repair_main(int argc, char **argv);
+ * @return 0 on success, SWUPD_THIRD_PARTY_REPO_EXISTS if it already exists
+ * and negative numbers on errors
+ */
+extern enum swupd_code third_party_add_main(int argc, char **argv);
```
I had mentioned this in a previous comment but I could not find it anymore, but you cannot use `enum swupd_code` and return negative numbers on errors. The whole purpose of using `enum soup_code` is to guarantee the function is going to return SWUPD_OK or SWUPD_NO if successful, and any other error within the swupd_exit_codes on errors.

castulo replied:
found the comment I'd made, here it is:

"All these functions are defined as returning a value of enum swupd_code, which at the moment are positive values from 0 to 36.

When defining functions within the swupd code it is possible to define functions that return any kind of value, boolean, positive numbers, negative numbers, etc. However whenever swupd exits, it must always exit with a positive value contained in swupd_code. In other words all functions that return values different than swupd_code can only be used internally, and whatever function that is going to propagate its return code all the way to the user has to return a value defined within swupd_code.

The whole purpose of using the enum swupd_code as return type is to "certify" in a way that a certain function complies with only returning positive values within the range of swupd_code, and this way we know it is safe to propagate its return code all the way to the user. So by this definition it makes no sense to define a function that supposedly complies with only returning values within swupd_code range, but arbitrarily returning negative numbers on errors. If a function returns a swupd_code then it must return SWUPD_OK or SWUPD_NO in a successful operation, and an appropriate error from swupd_code on failures."

karthikprabhuvinod replied:
Sorry about missing that. Will correct it in next push
